src/core/qgsvectorlayerutils.cpp

```diff
@@ -418,7 +418,7 @@ QgsFeature QgsVectorLayerUtils::createFeature( const QgsVectorLayer *layer, cons
 
     // 4. passed attribute value
     // note - deliberately not using else if!
-    if ( !v.isValid() && attributes.contains( idx ) )
+    if ( v.isValid() && attributes.contains( idx ) && !layer->primaryKeyAttributes().contains(idx) )
```

Comment:
I believe it should be:
```
if ( attributes.contains( idx ) && !layer->primaryKeyAttributes().contains(idx) )
```
otherwise you would skip the case in which `v` is not valid and we have an attribute.

Reply:
Agree! Will fix that!